### PR TITLE
cop: Support CASE_WHEN

### PR DESCRIPTION
**Pull request by breeswish**
Signed-off-by: Breezewish <breezewish@pingcap.com>

## What have you changed? (mandatory)

This PR adds RPN function CASE_WHEN().

## What are the type of the changes? (mandatory)

- Improvement (change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)

Unit test.

### CODE REVIEW

**Comment by lonng** [comment]
Please add some comments for `raw_arg`.

**Comment by breeswish** [comment]
@lonng Looks like `case_when` is a good enough example?

File: components/cop_codegen/src/rpn_function.rs [hunk 1]
```diff
@@ -279,6 +297,97 @@ impl VargsRpnFn {
...
+                ) -> crate::coprocessor::Result<crate::coprocessor::codec::data_type::VectorValue> #where_clause {
+                    crate::coprocessor::dag::rpn_expr::function::RAW_VARG_PARAM_BUF.with(|mut vargs_buf| {
+                        let mut vargs_buf = vargs_buf.borrow_mut();
+                        let args_len = args.len();
+                        let mut result = Vec::with_capacity(output_rows);
+                        for row_index in 0..output_rows {
+                            vargs_buf.clear();
+                            for arg_index in 0..args_len {
+                                let scalar_arg = args[arg_index].get_logical_scalar_ref(row_index);
+                                let scalar_arg = unsafe {
```
Comment (sticnarf) [inline components/cop_codegen/src/rpn_function.rs]: Since we still need unsafe, why not reuse `VARG_PARAM_BUF`?

Comment (sticnarf) [reply components/cop_codegen/src/rpn_function.rs]: Ah, got it, its size does not match...

File: src/coprocessor/dag/rpn_expr/impl_control.rs [hunk 2]
```diff
@@ -14,8 +14,30 @@ fn if_null<T: Evaluable>(lhs: &Option<T>, rhs: &Option<T>) -> Result<Option<T>>
     Ok(rhs.clone())
 }
 
+#[rpn_fn(raw_varg)]
+#[inline]
+pub fn case_when<T: Evaluable>(args: &[ScalarValueRef<'_>]) -> Result<Option<T>> {
+    for chunk in args.chunks(2) {
+        if chunk.len() == 1 {
+            // else statement
+            // TODO: Must verify type
```
Comment (sticnarf) [inline src/coprocessor/dag/rpn_expr/impl_control.rs]: With this concern, can we land this PR now? An invalid request just makes it panic.

Comment (breeswish) [reply src/coprocessor/dag/rpn_expr/impl_control.rs]: Invalid requests make other RPN function (including fixed args ones) panic already 😅 For example, pass in an `Int` while the function accept `Real`.

Comment (breeswish) [reply src/coprocessor/dag/rpn_expr/impl_control.rs]: Invalid requests make other RPN function (including fixed args ones) panic already 😅 For example, pass in an `Int` while the function accept `Real`.

Comment (breeswish) [reply src/coprocessor/dag/rpn_expr/impl_control.rs]: A fully functional validation will be resolved by https://github.com/tikv/tikv/pull/5033

File: src/coprocessor/dag/rpn_expr/impl_control.rs [hunk 3]
```diff
@@ -14,8 +14,30 @@ fn if_null<T: Evaluable>(lhs: &Option<T>, rhs: &Option<T>) -> Result<Option<T>>
...
+            let ret: &Option<T> = Evaluable::borrow_scalar_value_ref(&chunk[0]);
+            return Ok(ret.clone());
+        }
+        let cond: &Option<Int> = Evaluable::borrow_scalar_value_ref(&chunk[0]);
+        if cond == &Some(1) {
```
Comment (sticnarf) [inline src/coprocessor/dag/rpn_expr/impl_control.rs]: To be consistent with TiDB, this should be `cond.unwrap_or(0) != 0`?

**Pull request merged**